manta-trusted-setup/src/ceremony/signature.rs

```diff
@@ -0,0 +1,338 @@
+// Copyright 2019-2022 Manta Network.
+// This file is part of manta-rs.
+//
+// manta-rs is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// manta-rs is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with manta-rs.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Trusted Setup Ceremony Signatures
+
+use alloc::vec::Vec;
+use manta_crypto::{
+    dalek::ed25519::{Ed25519, SignatureError},
+    signature,
+};
+use manta_util::AsBytes;
+
+#[cfg(feature = "serde")]
+use manta_util::serde::{Deserialize, Serialize};
+
+/// Nonce
+pub trait Nonce: Default + PartialEq {
+    /// Increments the current nonce by one.
+    fn increment(&mut self);
+
+    /// Checks if the current nonce is valid.
+    fn is_valid(&self) -> bool;
+
+    /// Checks that `self` and `rhs` are valid and are both equal.
+    #[inline]
+    fn matches(&self, rhs: &Self) -> bool {
+        self.is_valid() && rhs.is_valid() && self == rhs
+    }
+}
+
+impl Nonce for u64 {
+    #[inline]
+    fn increment(&mut self) {
+        *self = self.saturating_add(1);
+    }
+
+    #[inline]
+    fn is_valid(&self) -> bool {
+        *self != Self::MAX
+    }
+}
+
+/// Message with Nonce
+#[derive(Clone, Debug, Default, Eq, Hash, PartialEq)]
+pub struct RawMessage<N> {
+    /// Nonce
+    nonce: N,
+
+    /// Encoded Message
+    encoded_message: Vec<u8>,
+}
+
+impl<N> RawMessage<N> {
+    /// Builds a new [`RawMessage`] from `nonce` and `encoded_message`.
+    #[inline]
+    pub fn new(nonce: N, encoded_message: Vec<u8>) -> Self {
+        Self {
+            nonce,
+            encoded_message,
+        }
+    }
+}
+
+impl<N> AsBytes for RawMessage<N>
+where
+    N: AsBytes,
+{
+    #[inline]
+    fn as_bytes(&self) -> Vec<u8> {
+        let mut bytes = self.nonce.as_bytes();
+        bytes.extend_from_slice(&self.encoded_message);
+        bytes
+    }
+}
+
+/// Signature Scheme
+pub trait SignatureScheme:
+    Default
+    + signature::Sign<Message = RawMessage<Self::Nonce>, Randomness = ()>
+    + signature::Verify<Verification = Result<(), Self::Error>>
+{
+    /// Message Nonce Type
+    type Nonce: Clone + Nonce;
+
+    /// Verification Error Type
+    type Error;
+}
+
+impl<N> SignatureScheme for Ed25519<RawMessage<N>>
+where
+    N: AsBytes + Clone + Default + Nonce,
+{
+    type Nonce = N;
+    type Error = SignatureError;
+}
+
+/// Signs the `message` with the `nonce` attached using the `signing_key`.
+#[cfg(feature = "bincode")]
```

Comment:
Typically for signature schemes, we want a binary encoding rather than a self-describing format like JSON, it's just a little simpler. Because we don't have a formal versioning scheme right now it's not a big deal but we can formalize this in subsequent versions.